jupyterhub/spawner.py

```diff
@@ -119,50 +165,93 @@ def options_from_form(self, form_data):
         'LANG',
         'LC_ALL',
     ],
-        help="Whitelist of environment variables for the subprocess to inherit"
+        help="""
+        Whitelist of environment variables for the single-user server to inherit from the JupyterHub process.
+
+        This whitelist is used to ensure that sensitive information in the JupyterHub process's environment
+        (such as `CONFIGPROXY_AUTH_TOKEN`) is not passed to the single-user server's process.
+        """
     ).tag(config=True)
+
     env = Dict(help="""Deprecated: use Spawner.get_env or Spawner.environment
-    
+
     - extend Spawner.get_env for adding required env in Spawner subclasses
     - Spawner.environment for config-specified env
     """)
-    
+
     environment = Dict(
-        help="""Environment variables to load for the Spawner.
+        help="""
+        Extra environment variables to set for the single-user server's process.
+
+        Environment variables that end up in the single-user server's process come from 3 sources:
+          - This `environment` configurable
+          - The JupyterHub process' environment variables that are whitelisted in `env_keep`
+          - Variables to establish contact between the single-user notebook and the hub (such as JPY_API_TOKEN)
 
-        Value could be a string or a callable. If it is a callable, it will
-        be called with one parameter, which will be the instance of the spawner
-        in use. It should quickly (without doing much blocking operations) return
-        a string that will be used as the value for the environment variable.
+        The `enviornment` configurable should be set by JupyterHub administrators to add
+        installation specific environment variables. It is a dict where the key is the name of the environment
+        variable, and the value can be a string or a callable. If it is a callable, it will be called
+        with one parameter (the spawner instance), and should return a string fairly quickly (no blocking
+        operations please!).
+
+        Note that the spawner class' interface is not guaranteed to be exactly same across upgrades,
+        so if you are using the callable take care to verify it continues to work after upgrades!
         """
     ).tag(config=True)
-    
-    cmd = Command(['jupyterhub-singleuser'],
-        help="""The command used for starting notebooks."""
+
+    # FIXME: Add info about shell expansion here.
+    cmd = Command(
+        ['jupyterhub-singleuser'],
+        help="""
+        The command used for starting the single-user server.
+
+        You can provide this as either a list or a string. Note that this should only contain
```

Comment:
Provide either a string or a list containing the path to the startup script command. Extra arguments, other than this path, should be provided via...